Assets/Scripts/Player/PlayerVerticalMovement.cs

```diff
@@ -0,0 +1,142 @@
+    private float currentStateStartTime;                    // when we entered the current state
+
+    private bool jumpPressed = false;           // current state of player jump input
+
+    private float verticalVelocity = 0.0f;      // current vertical velocity (direction matters)
+    private float gravity = -0.3f;              // current gravity value
+
+    /// <summary>Update player input for jumping.</summary
+    public void UpdateJumpPressed(bool playerInput) { jumpPressed = playerInput; }
+
+    /// <summary>Update the current vertical velocity and return its signed-value.</summary>
+    /// <param name="isGrounded">whether the player is touching the ground</param>
+    /// <param name="defaultGravity">default gravity when not jumping / floating</param>
+    /// <param name="hoverHeight">how high to hover</param>
+    /// <param name="hoverRiseTime">how long jump to hoverHeight takes</param>
+    /// <param name="maxHoverTime">how long to hover before landing</param>
+    /// <returns></returns>
+    public float UpdateVelocity(bool isGrounded, float defaultGravity, float hoverHeight, float hoverRiseTime, float maxHoverTime)
+    {
+        AdvanceJumpStateMachine(isGrounded, maxHoverTime);
+        SetGravityForJumpState(defaultGravity, hoverHeight, hoverRiseTime, maxHoverTime);
+
+        // update velocity with current gravity
+        if (isGrounded && verticalVelocity < 0.0f) verticalVelocity = 0.0f;
+        else verticalVelocity += gravity * Time.deltaTime;
+        return verticalVelocity;
+    }
+
+
+    /// <summary>Updates the current jump-state per user input and state.</summary>
+    /// <param name="isGrounded">Whether the character is currently touching the ground.</param>
+    /// <param name="maxHoverTime">Max amount of time that can be spent in Hover state before falling.</param>
+    private void AdvanceJumpStateMachine(bool isGrounded, float maxHoverTime)
+    {
+        JumpState nextState = currentState;
+
+        switch (currentState)
+        {
+            case JumpState.DEFAULT:
+                /// progress from default state to idle       
+                nextState = JumpState.IDLE;
+                break;
+
+            case JumpState.IDLE:
+                /// can jump from idle, if touching ground
+                if (jumpPressed && isGrounded) { nextState = JumpState.JUMPING; }
+                break;
+
+            case JumpState.JUMPING:
+                // can release button to fall faster
+                if (!jumpPressed) { nextState = JumpState.RELEASED; }
+                // transition to FLOATING at peak of jump 
+                else if (verticalVelocity <= 0.0f) { nextState = JumpState.FLOATING; }
+                break;
+
+            case JumpState.FLOATING:
+                // can release button to fall faster
+                // OR exceed maxHovertime
+                if (!jumpPressed || (Time.time - currentStateStartTime) > maxHoverTime) { nextState = JumpState.RELEASED; }
```

Comment:
we could take a fall-speed based approach to it too - I just designed it this way because it made it easier for me to think about "how far" you can float for level design.

Like, if I know my max horizontal speed is 10 units/sec and my float time is 3 seconds then I know I can design a gap 30 units wide before the player can't cross it. But if I had to design in terms of acceleration and gravity values, I think it would get pretty wonky pretty quick! 

We could also have items increase the hover time - adding a second or two to that number is pretty ez